tests/trainer/test_trainer.py

```diff
@@ -835,6 +836,41 @@ def test_number_of_steps_in_training_with_ipex(self):
             train_output = trainer.train()
             self.assertEqual(train_output.global_step, 10)
 
+    def test_neftune(self):
+        config = GPT2Config(vocab_size=100, n_positions=128, n_embd=32, n_layer=3, n_head=4)
+        tiny_gpt2 = GPT2LMHeadModel(config)
+        x = torch.randint(0, 100, (128,))
+        train_dataset = RepeatDataset(x)
+
+        # Trainer without inf/nan filter
+        args = TrainingArguments(
+            "./test", learning_rate=1e9, logging_steps=5, logging_nan_inf_filter=False, neftune_noise_alpha=0.4
+        )
+        trainer = Trainer(tiny_gpt2, args, train_dataset=train_dataset)
+
+        trainer.model = trainer._activate_neftune(trainer.model)
+
+        dummy_input = torch.LongTensor([[1, 0, 1]]).to(torch_device)
+
+        emb1 = trainer.model.get_input_embeddings()(dummy_input)
+        emb2 = trainer.model.get_input_embeddings()(dummy_input)
+        self.assertFalse(torch.allclose(emb1, emb2), "Neftune noise is not applied!")
+
+        # redefine the model
+        tiny_gpt2 = GPT2LMHeadModel(config)
+        # Trainer without inf/nan filter
+        args = TrainingArguments(
+            "./test", learning_rate=1e9, logging_steps=5, logging_nan_inf_filter=False, neftune_noise_alpha=0.4
+        )
+        trainer = Trainer(tiny_gpt2, args, train_dataset=train_dataset)
+
+        # Check that it trains without errors
+        trainer.train()
+
+        # Make sure forward pass works fine
+        _ = trainer.model(torch.LongTensor([[1, 0, 1]]).to(torch_device))
+        self.assertTrue(len(trainer.model.get_input_embeddings()._forward_hooks) == 0)
```

Comment:
A check should be made that it's correctly disabled after training has finished

Reply:
the line 
```python
self.assertTrue(len(trainer.model.get_input_embeddings()._forward_hooks) == 0)
```
Should check if the forward hook as been correctly removed so I think all should be good here

Reply:
Note also that line is called right after training, so it should check that neftune is correctly disabled after training.

Reply:
Added slightly more elaborated test in [ca8f8c4](https://github.com/huggingface/transformers/pull/27141/commits/ca8f8c43f662f55e367bb035fa2a51abf78fc794)